### PR TITLE
Object detection scorer

### PR DESCRIPTION
**Pull request by IgorSusmelj**
### Changes
- Added the framework for computing scores for object detection predictions
- Add the scorer for "frequency-detection" which prefers images with many objects and diverse objects

### CODE REVIEW

File: lightly/active_learning/scorers/detection.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+from typing import *
+
+import numpy as np
+
+from lightly.active_learning.scorers.scorer import Scorer
+
+
+def _object_frequency(model_output: List[Dict],
+                      frequency_pentaly: float,
```
Comment (philippmwirth) [inline lightly/active_learning/scorers/detection.py:9]: "pentaly" ;)

**Review by philippmwirth — CHANGES_REQUESTED** [review]
Looks good, I have some minor feedback.

One thing I'd also like to discuss: We're requesting a rather specific format, I know that for example torchvision FasterRCNN provides a different output. Maybe we should already anticipate this and think about which is the most general type of boxes we can get -> make this the parent class and then have the others inherit from that. What do you think?

**Comment by IgorSusmelj** [comment]
> Looks good, I have some minor feedback.
> 
> One thing I'd also like to discuss: We're requesting a rather specific format, I know that for example torchvision FasterRCNN provides a different output. Maybe we should already anticipate this and think about which is the most general type of boxes we can get -> make this the parent class and then have the others inherit from that. What do you think?

I like the idea of being more flexible. But there are so many options in object detection. I feel the best way is to have a clear format we want but help the user get there. Maybe we could offer some helper functions to the user. E.g. to change bounding box representation (x1, y1, x2, y2) --> (x1, y1, w, h).
Do you think that would work?

File: lightly/active_learning/scorers/detection.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            model. For each sample we expect a dictionary containing `boxes`,
+            `object_probabilities` and `class_probabilities`.
+            `boxes` has shape [B, 4] for B found bounding boxes and the 4 values
+            x1, y1, x2, y2.
+            `object_probabilities` has shape [B] with a objectness value for each
+            bounding box.
+            `class_probabilities` has shape [B, C] for the B bounding boxes and
+            C classes. The class probabilities should all sum up to 1.
+        config:
+            A dictionary containing additional parameters for the scorers.
```
[the remark this answers is not in the document]
Comment (IgorSusmelj) [reply lightly/active_learning/scorers/detection.py:61]: I will add one

File: lightly/active_learning/scorers/detection.py (hunk 1 above)
Comment (IgorSusmelj) [reply lightly/active_learning/scorers/detection.py:9]: arrgh, will fix this :)

**Comment by philippmwirth** [comment]
> I like the idea of being more flexible. But there are so many options in object detection. I feel the best way is to have a clear format we want but help the user get there. Maybe we could offer some helper functions to the user. E.g. to change bounding box representation (x1, y1, x2, y2) --> (x1, y1, w, h).
> Do you think that would work?

I think there are many approaches we could take there:
- Bounding Box Class for each type of possible bbox and all provide the necessary functions for our scorer.
- Scorer Class for each type of possible bbox and all calculate the scores they can
- Helper functions which make it easy for the user to switch between formats

I would prefer the first one but maybe this should be covered in a different ticket.

**Comment by IgorSusmelj** [comment]
> > I like the idea of being more flexible. But there are so many options in object detection. I feel the best way is to have a clear format we want but help the user get there. Maybe we could offer some helper functions to the user. E.g. to change bounding box representation (x1, y1, x2, y2) --> (x1, y1, w, h).
> > Do you think that would work?
> 
> I think there are many approaches we could take there:
> 
> * Bounding Box Class for each type of possible bbox and all provide the necessary functions for our scorer.
> * Scorer Class for each type of possible bbox and all calculate the scores they can
> * Helper functions which make it easy for the user to switch between formats
> 
> I would prefer the first one but maybe this should be covered in a different ticket.

Let's discuss this further in another issue. I'm also not sure where to put it. To `utils`? I would prefer 1 or 3. I could imagine that the 3rd option could be the easiest to implement. A few methods to transform the bounding boxes. I'm not sure a class would help.

**Comment by IgorSusmelj** [comment]
I created an issue for the further discussion: https://github.com/lightly-ai/lightly/issues/253

**Review by philippmwirth — APPROVED** [review]

**Pull request merged**